Improve compatibility between pipeline & cluster

Cecilia-Sensalari opened the pull request:
- Removed $NSLOTS in main.nf, which is defined only by SGE; in its place, $task.cpus is used
- Added Nextflow directives in configuration files such as cpus, penv, memory
- Updated Nextflow configuration file description in documentation

lohausr commented on doc/source/template_docker_nextflow.config:
```diff
@@ -18,11 +17,17 @@ process {
     // It is advised to provide higher computational power to wgdParalogs and wgdOrthologs 
     withName: 'wgdParalogs' {
-        clusterOptions = ''  // e.g. '-pe serial 10 -l h_vmem=2G'
+        cpus = ''            // e.g. '8'
+        penv = ''            // e.g. 'serial'
+        memory = ''          // e.g. '16GB'
```
Do cpus and memory values need to be in quotes or does it matter?

lohausr commented on doc/source/template_docker_nextflow.config:
```diff
@@ -18,11 +17,17 @@ process {
     
     withName: 'wgdOrthologs' {
-        clusterOptions = ''  // e.g. '-pe serial 10 -l h_vmem=2G'
+        cpus = ''            // e.g. '8'
+        penv = ''            // e.g. 'serial'
+        memory = ''          // e.g. '16GB'
```
Do cpus and memory values need to be in quotes or does it matter?

lohausr replied:
Wouldn't it make sense to also add at least the `memory` setting also for the other processes below and not just `clusterOptions`?

Cecilia-Sensalari replied:
Agreed!

lohausr commented on doc/source/template_nocontainer_nextflow.config:
```diff
@@ -12,12 +12,18 @@ process {
     // It is advised to provide higher computational power to wgdParalogs and wgdOrthologs 
     withName: 'wgdParalogs' {
-        clusterOptions = ''  // e.g. '-pe serial 10 -l h_vmem=2G'
+        cpus = ''            // e.g. '8'
+        penv = ''            // e.g. 'serial'
+        memory = ''          // e.g. '16GB'
```
See comments on other template.

lohausr commented on doc/source/template_nocontainer_nextflow.config:
```diff
@@ -12,12 +12,18 @@ process {
+        penv = ''            // e.g. 'serial'
+        memory = ''          // e.g. '16GB'
+        clusterOptions = ''  // e.g. '-l h_vmem=2G'
         beforeScript = ''    // e.g. 'module load python paml muscle mafft prank fasttree blast mcl i-adhore'
```
I don't think we offer options (yet) to specify MAFFT or PRANK instead of MUSCLE, so better not list them here?

Cecilia-Sensalari replied:
Agreed!

lohausr commented on doc/source/template_nocontainer_nextflow.config:
```diff
@@ -12,12 +12,18 @@ process {
+        penv = ''            // e.g. 'serial'
+        memory = ''          // e.g. '16GB'
+        clusterOptions = ''  // e.g. '-l h_vmem=2G'
         beforeScript =  ''   // e.g. 'module load python paml muscle mafft prank blast mcl'
```
See above for mafft and prank, also wgd ortholog analyses don't need mcl.

Cecilia-Sensalari replied:
Agreed!

lohausr commented on doc/source/template_nocontainer_nextflow.config:
```diff
@@ -40,8 +46,3 @@ process {
-// params {
-//     nThreadsParalogs = 
-//     nThreadsOrthologs = 
-// }
```
As far as I can see this template still has a `container` setting.

Cecilia-Sensalari replied:
Not needed indeed, thanks.

lohausr commented on doc/source/template_singularity_nextflow.config:
```diff
@@ -19,11 +18,17 @@ process {
     // It is advised to provide higher computational power to wgdParalogs and wgdOrthologs 
     withName: 'wgdParalogs' {
-        clusterOptions = ''  // e.g. '-pe serial 10 -l h_vmem=2G'
+        cpus = ''            // e.g. '8'
+        penv = ''            // e.g. 'serial'
+        memory = ''          // e.g. '16GB'
```
See comments on other template.

lohausr commented on example/custom_nextflow.config:
```diff
@@ -25,11 +23,17 @@ process {
     // It is advised to provide higher computational power to wgdParalogs and wgdOrthologs 
     withName: 'wgdParalogs' {
-        clusterOptions = ''  // e.g. '-pe serial 10 -l h_vmem=2G'
+        cpus = ''            // e.g. '8'
+        penv = ''            // e.g. 'serial'
+        memory = ''          // e.g. '16GB'
```
See comments on other templates.

lohausr commented on test/custom_nextflow.config:
```diff
@@ -25,19 +23,25 @@ process {
     // It is advised to provide higher computational power to wgdParalogs and wgdOrthologs 
     withName: 'wgdParalogs' {
-        clusterOptions = ''  // e.g. '-pe serial 10 -l h_vmem=2G'
+        cpus = ''            // e.g. '8'
+        penv = ''            // e.g. 'serial'
+        memory = ''          // e.g. '16GB'
```
See comments on other templates.

lohausr commented on main.nf:
```diff
@@ -481,17 +481,10 @@ process wgdParalogs {
     echo "NF internal work directory for [wgdParalogs] process:\n\$processDir\n" >> $logs_folder/wgd_paralogs.log
 
-    if [ -z \${NSLOTS+x} ]; then
-        nThreads=$params.nThreadsParalogs
```
Did you also remove the parameter `nThreadsParalogs`? Or do we still need it for something else?

lohausr commented on main.nf:
```diff
@@ -481,17 +481,10 @@ process wgdParalogs {
-        echo "Found \\\$NSLOTS = \$NSLOTS\n" >> $logs_folder/wgd_paralogs.log
-        echo "[$species] Found \\\$NSLOTS = \$NSLOTS -> using \${nThreads} threads"
-    fi
+    echo "[$species] Using ${task.cpus} thread(s)"
+    echo "Using ${task.cpus} thread(s)\n">> $logs_folder/wgd_paralogs.log
```
Better to write to logs first.

lohausr commented on main.nf:
```diff
@@ -481,17 +481,10 @@ process wgdParalogs {
+    echo "Using ${task.cpus} thread(s)\n">> $logs_folder/wgd_paralogs.log
 
-    ksrates paralogs-ks ${config} --n-threads=\$nThreads >> $logs_folder/wgd_paralogs.log 2>&1
+    ksrates paralogs-ks ${config} --n-threads=${task.cpus} >> $logs_folder/wgd_paralogs.log 2>&1
```
Did you test yet whether this works also locally?

lohausr commented on main.nf:
```diff
@@ -554,17 +547,10 @@ process wgdOrthologs {
     echo "NF internal work directory for [wgdOrthologs (${task.index})] process:\n\$processDir\n" > $logs_folder/wgd_orthologs_${species1}_${species2}.log
 
-    if [ -z \${NSLOTS+x} ]; then
-        nThreads=$params.nThreadsOrthologs
```
Did you also remove the parameter `nThreadsOrthologs`? Or do we still need it for something else?

lohausr commented on main.nf:
```diff
@@ -554,17 +547,10 @@ process wgdOrthologs {
-        echo "Found \\\$NSLOTS = \$NSLOTS\n" >> $logs_folder/wgd_orthologs_${species1}_${species2}.log
-        echo "[$species1 – $species2] Found \\\$NSLOTS = \$NSLOTS -> using \${nThreads} threads"
-    fi
+    echo "[$species1 – $species2] Using ${task.cpus} thread(s)"
+    echo "Using ${task.cpus} thread(s)\n">> $logs_folder/wgd_orthologs_${species1}_${species2}.log
```
Better to write to logs first.

lohausr commented on main.nf:
```diff
@@ -554,17 +547,10 @@ process wgdOrthologs {
+    echo "Using ${task.cpus} thread(s)\n">> $logs_folder/wgd_orthologs_${species1}_${species2}.log
 
-    ksrates orthologs-ks ${config} $species1 $species2 --n-threads=\$nThreads >> $logs_folder/wgd_orthologs_${species1}_${species2}.log 2>&1
+    ksrates orthologs-ks ${config} $species1 $species2 --n-threads=${task.cpus} >> $logs_folder/wgd_orthologs_${species1}_${species2}.log 2>&1
```
Did you test yet whether this works also locally?

lohausr reviewed (changes requested):
See comments.

lohausr commented on doc/source/configuration.rst:
```diff
@@ -170,9 +169,12 @@ The Nextflow configuration file is used to configure various settings for the *k
+      Cluster settings can be tailored on your own machine type through the use of Nextflow directives (for a complete list see the `Nextflow documentation <https://www.nextflow.io/docs/latest/process.html#process-directives>`__), such as:
     
-    	* **clusterOption** any native configuration option accepted by your cluster submit command. You can use it to request non-standard resources or use settings that are specific to your cluster and not supported out of the box by Nextflow.
+        * **cpus** sets the number of CPUs/cores/slots/threads, e.g. ``8`` [Default if not set: 1]
```
Maybe we should stress more that this directive is mandatory or needs to be set if one wants more than 1 CPU?

Cecilia-Sensalari replied:
I can add a short sentence to suggest to use multiple threads for the two wgd processes.

lohausr replied:
The comment was more about that it's required to use this directive to set more than 1 CPU (as opposed to e.g. `-pe serial 8`).

Cecilia-Sensalari replied:
I tried it out again: cpus can be without quote, but memory and time need quotes (perhaps because they're a string such as 3GB or 1h, while cpus is a number). I'll remove the quotes from cpus!

Cecilia-Sensalari commented on doc/source/configuration.rst:
```diff
@@ -139,24 +137,25 @@ The Nextflow configuration file is used to configure various settings for the *k
         container = ''
 
         withName: 'processName' {
+            cpus = ''
+            penv = ''
+            memory = ''
```
Yes, they need to be in quotes or else they are not recognized.

lohausr commented on doc/source/faqs.rst:
```diff
@@ -72,6 +72,12 @@ The Nextflow ``wgd_paralogs.log`` file reports details over the *K*:sub:`S` anal
 .. -----------------------------------------
 
 
+How to solve ``failed to reserve page summary memory`` error? 
+-------------------------------------------------------------
+
+Some clusters might not be compatible with certain Nextflow built-in directives in the Nextflow configuration file. For example, running the pipeline after having set the ``memory`` directive could return the error message ``failed to reserve page summary memory`` because the cluster expects the memory usage to be expressed per CPU instead of as overall amount. In these cases we advise to set the ...
+
+
```
```suggestion
I am getting a ``failed to reserve page summary memory`` error?
-------------------------------------------------------------

Some cluster configurations may not be fully compatible with certain built-in Nextflow process directives used in the Nextflow configuration file. For example, on some cluster configurations the *ksrates* pipeline fails with an ``failed to reserve page summary memory`` error when using the built-in ``memory`` directive to define how much memory a process requires. In these cases it is likely that instead the ``clusterOptions`` process directive (see the `Nextflow documentation <https://www.nextflow.io/docs/latest/process.html#clusteroptions>`__) needs to be used to define configuration settings specific to your cluster. We advise to contact your cluster administrator about these.
```

lohausr replied:
Since this is not a general error but specific to Nextflow, I would move it up into the Nextflow section of the FAQ.

Cecilia-Sensalari replied:
Edits committed and I moved the paragraph to the Nextflow section.

Cecilia-Sensalari commented:
UPDATES
Nextflow configuration files:
- Add `cpus` directive under `executor` for when `executor` is set to `local`: it limits the number of CPUs available for the executor (e.g. Docker contains would otherwise use as much resources as they can).
- Add `queueSize` directive under `executor` for when the pipeline is run on a cluster, to limit the number of processes simultaneously handled by the executor (handy when the user can't require too many CPUs at the same time).
- Removed `sequential`, `nThreadsParalogs`, `nThreadsOrthologs` and unnecessary `maxForks` parameters in main.nf
- Docs updated with description of `queueSize` and `cpus`.

Cecilia-Sensalari commented on doc/source/configuration.rst:
```diff
@@ -170,9 +169,12 @@ The Nextflow configuration file is used to configure various settings for the *k
-    	* **clusterOption** any native configuration option accepted by your cluster submit command. You can use it to request non-standard resources or use settings that are specific to your cluster and not supported out of the box by Nextflow.
+        * **cpus** sets the number of CPUs/cores/slots/threads, e.g. ``8`` [Default if not set: 1]
+    	* **penv** defines the parallel environment to be used when submitting a parallel task to the SGE resource manager
+        * **memory** sets how much memory the process is allowed to use, e.g. ``16GB``. Note that, in case of multiple CPUs, this is the overall amount of memory dedicated to the CPUs, not the amount of memory *per* CPU.
```
Okay, I'm removing the second part (starting from "Note that").

Cecilia-Sensalari commented:
Updates:
- Docs considers now Nextflow configuration file as a required input and not anymore as optional
- Add sentence about setting queueSize to 1 to have sequential workflow

lohausr commented on doc/source/template_docker_nextflow.config:
```diff
@@ -8,7 +8,11 @@ docker {
 // CLUSTER SETTINGS:
-executor.name = ''  // set cluster type, e.g. sge
+executor {
+    name = ''                   // e.g. sge, local
```
```suggestion
    name = ''                   // e.g. 'sge', 'local'
```

lohausr commented on doc/source/template_docker_nextflow.config:
```diff
@@ -8,7 +8,11 @@ docker {
-executor.name = ''  // set cluster type, e.g. sge
+executor {
+    name = ''                   // e.g. sge, local
+    // queueSize =              // e.g. 4, to limit the number of processes simultaneously handled
```
```suggestion
    // queueSize =              // e.g. 4, to limit the number of processes handled in parallel
```

lohausr commented on doc/source/template_docker_nextflow.config:
```diff
@@ -8,7 +8,11 @@ docker {
+executor {
+    name = ''                   // e.g. sge, local
+    // queueSize =              // e.g. 4, to limit the number of processes simultaneously handled
+    // cpus =                   // e.g. 8, to limit the CPU resources when using local executor
```
```suggestion
    // cpus =                   // e.g. 3, to limit the CPU resources when using the 'local' executor
```

lohausr replied:
8 seemed like a high number to use as an example limit for a local computer, not many laptops have more than 8 CPUs/cores.

lohausr commented on doc/source/template_docker_nextflow.config:
```diff
@@ -17,32 +21,32 @@ process {
-        penv = ''            // e.g. 'serial'
-        memory = ''          // e.g. '16GB'
-        clusterOptions = ''  // e.g. '-l h_vmem=2G'
+        cpus =                  // e.g. 8
```
Maybe change the comment to something like "e.g. 8 on a cluster or 2 on local computer"? Same for below.

Cecilia-Sensalari replied:
Agreed!

lohausr commented on doc/source/template_nocontainer_nextflow.config:
```diff
@@ -3,44 +3,48 @@
 // CLUSTER SETTINGS:
-executor.name = ''  // set cluster type, e.g. sge
+executor {
+    name = ''                   // e.g. sge, local
+    // queueSize =              // e.g. 4, to limit the number of processes simultaneously handled
+    // cpus =                   // e.g. 8, to limit the CPU resources when using local executor
```
```suggestion
    name = ''                   // e.g. 'sge', 'local'
    // queueSize =              // e.g. 4, to limit the number of processes handled in parallel
    // cpus =                   // e.g. 3, to limit the CPU resources when using the 'local' executor
```

lohausr commented on doc/source/template_singularity_nextflow.config:
```diff
@@ -10,40 +10,44 @@ singularity {
 // CLUSTER SETTINGS:
-executor.name = ''  // set cluster type, e.g. sge
+executor {
+    name = ''                   // e.g. sge, local
+    // queueSize =              // e.g. 4, to limit the number of processes simultaneously handled
+    // cpus =                   // e.g. 8, to limit the CPU resources when using local executor
```
```suggestion
    name = ''                   // e.g. 'sge', 'local'
    // queueSize =              // e.g. 4, to limit the number of processes handled in parallel
    // cpus =                   // e.g. 3, to limit the CPU resources when using the 'local' executor
```

lohausr commented on doc/source/usage.rst:
```diff
@@ -16,14 +16,14 @@ Run example case as a Nextflow pipeline (recommended)
 2.  Prepare the configuration files.
 
-    The directory already contains a pre-filled *ksrates configuration file* (``config_elaeis.txt``) and a *Nextflow configuration file* template (``custom_nextflow.config``). If running on a cluster or within a container, please fill in the template as described in the :ref:`nextflow_config_section` section.
+    The directory already contains a pre-filled *ksrates configuration file* (``config_elaeis.txt``) and a *Nextflow configuration file* template (``custom_nextflow.config``) to be filled in as described in the :ref:`nextflow_config_section` section.
```
Would it make sense to rename the custom_nextflow.config to simply the default nextflow.config? Otherwise we should add the "-c custom_nextflow.config" to all nextflow command lines everywhere in the documentation (e.g. also in Installation -> Testing your installation).

Cecilia-Sensalari replied:
Now that it is a "mandatory" file it makes much sense to do that. By renaming it just nextflow.config it will be automatically detected and used, and we can drop the "-c" flag in the launching command.
Shall I delete it from everywhere then? I'll just note down in the Usage that you can explicitly give the path with "-c" if you want to call your file another way or if you want to give a specific path.

lohausr replied:
Sounds good to me, just make sure that it's still described that it is needed and should be called nextflow.config.

lohausr commented on doc/source/usage.rst:
```diff
@@ -36,7 +36,7 @@ The *ksrates* pipeline can be automatically run through Nextflow with a few prep
         nextflow run VIB-PSB/ksrates --config ./config_elaeis.txt -c ./custom_nextflow.config
 
-    The ``--config`` option takes the *ksrates configuration file*, while ``-c`` takes the optional *Nextflow configuration file*. If the Nextflow-reserved ``nextflow.config`` name is used, this latter file is automatically recognized without explicitly calling it in the command line.
+    The ``--config`` option takes the *ksrates configuration file*, while ``-c`` takes the *Nextflow configuration file*. If the Nextflow-reserved ``nextflow.config`` name is used, this latter file is automatically recognized without explicitly calling it in the command line.
```
```suggestion
    The ``--config`` option takes the *ksrates configuration file*, while ``-c`` takes the *Nextflow configuration file*. If the Nextflow-reserved ``nextflow.config``file name is used for the latter, the file is automatically recognized without having explicitly configure it with the ``-c`` option.
```

lohausr commented on example/custom_nextflow.config:
```diff
@@ -8,47 +8,51 @@ singularity {
 // CLUSTER SETTINGS:
-executor.name = ''  // e.g. sge
+executor {
+    name = ''                   // e.g. sge, local
+    // queueSize =              // e.g. 4, to limit the number of processes simultaneously handled
+    // cpus =                   // e.g. 8, to limit the CPU resources when using local executor
```
```suggestion
    name = ''                   // e.g. 'sge', 'local'
    // queueSize =              // e.g. 4, to limit the number of processes handled in parallel
    // cpus =                   // e.g. 3, to limit the CPU resources when using the 'local' executor
```

lohausr commented on example/custom_nextflow.config:
```diff
@@ -8,47 +8,51 @@ singularity {
-    container = 'docker://vibpsb/ksrates:latest'      // to build a Singularity container
-//  container = 'vibpsb/ksrates:latest'               // to build a Docker container
+    // Choose a container type (not needed if all dependencies are locally installed)
+    container = 'docker://vibpsb/ksrates:latest'        // to build a Singularity container
+    // container = 'vibpsb/ksrates:latest'              // to build a Docker container
```
```suggestion
    container = 'docker://vibpsb/ksrates:latest'        // to use the ksrates Singularity container
    // container = 'vibpsb/ksrates:latest'              // to use the ksrates Docker container
```

lohausr commented on test/custom_nextflow.config:
```diff
@@ -8,47 +8,50 @@ singularity {
 // CLUSTER SETTINGS:
-executor.name = ''  // e.g. sge
+executor {
+    name = 'local'              // e.g. sge
+    cpus = 2                    // to limit the CPU resources when using local executor
```
```suggestion
    name = 'local'              // or e.g. 'sge' when using a computer cluster
    cpus = 3                    // to limit the CPU resources when using the 'local' executor
```

lohausr commented on test/custom_nextflow.config:
```diff
@@ -8,47 +8,50 @@ singularity {
-    container = 'docker://vibpsb/ksrates:latest'      // to build a Singularity container
-//  container = 'vibpsb/ksrates:latest'               // to build a Docker container
+    // Choose a container type (not needed if all dependencies are locally installed)
+    container = 'docker://vibpsb/ksrates:latest'        // to build a Singularity container
+    // container = 'vibpsb/ksrates:latest'              // to build a Docker container
```
```suggestion
    container = 'docker://vibpsb/ksrates:latest'        // to use the ksrates Singularity container
    // container = 'vibpsb/ksrates:latest'              // to use the ksrates Docker container
```

lohausr commented on test/custom_nextflow.config:
```diff
@@ -8,47 +8,50 @@ singularity {
+    // enabled = true
+// }
 
 // CLUSTER SETTINGS:
```
```suggestion
// EXECUTOR SETTINGS:
```

lohausr commented on test/custom_nextflow.config:
```diff
@@ -8,47 +8,50 @@ singularity {
-        memory = ''          // e.g. '16GB'
-        clusterOptions = ''  // e.g. '-l h_vmem=2G'
+        cpus = 1                // e.g. 8
+        penv = ''               // e.g. 'serial'
+        memory = ''             // e.g. '16GB'
+        clusterOptions = ''     // e.g. '-l h_vmem=2G'
```
What happens is these are set to an empty values/string? Should we maybe comment these out since this seems to be pre-configured for a local computer? Same for below.

Cecilia-Sensalari replied:
When I ran it, it didn't complain about these fields being set to empty strings (so I guess it ignores them). But I can comment "penv", "memory" and "clusterOptions" out to be more consistent with the "local" pre-setting.

Cecilia-Sensalari replied:
Tried that and there is problem: the withName section complains that there are no directives within it! So, either I comment out the whole withName section like below or we leave it as it was. I'll go for the first for now, although the user will have to uncomment quite some stuff if they want to run the test on the cluster (which is however not necessary).
```
    // withName: 'paralogsAnalyses' {
    //     memory = ''             // e.g. '2GB'
    //     clusterOptions = ''     // e.g. '-l h_vmem=2G'
    // }
```

lohausr replied:
Ah, I think that's OK.

lohausr commented on example/custom_nextflow.config:
```diff
@@ -8,47 +8,51 @@ singularity {
+    // enabled = true
+// }
 
 // CLUSTER SETTINGS:
```
```suggestion
// EXECUTOR SETTINGS:
```

lohausr commented on doc/source/template_singularity_nextflow.config:
```diff
@@ -10,40 +10,44 @@ singularity {
 }
 
 // CLUSTER SETTINGS:
```
```suggestion
// EXECUTOR SETTINGS:
```

lohausr commented on doc/source/template_nocontainer_nextflow.config:
```diff
@@ -3,44 +3,48 @@
 
 
 // CLUSTER SETTINGS:
```
```suggestion
// EXECUTOR SETTINGS:
```

lohausr commented on doc/source/template_docker_nextflow.config:
```diff
@@ -8,7 +8,11 @@ docker {
 }
 
 // CLUSTER SETTINGS:
```
```suggestion
// EXECUTOR SETTINGS:
```

lohausr reviewed (changes requested):
See comments and changes. Readme.md also needs to be updated e.g. regarding always requiring the Nextflow configuration file.

Cecilia-Sensalari commented:
UPDATES:
- Accepted review edits
- NF configuration files for test and example dataset have been renamed `nextflow.config` and the docs explain how to use `-c` option
- Readme has been updated: Nextflow config file is now mandatory, the `-c` option is removed, the distinction between local and remote run has been removed and I added a "Prepare the configuration files" step (such as in usage.rst).

lohausr commented on README.md:
```diff
@@ -80,35 +80,17 @@ We briefly illustrate here how to run the *ksrates* Nextflow pipeline on the `te
+	   
+   The `test` directory already contains a pre-filled *ksrates* pipeline configuration file (`config_elaeis.txt`) for the oil palm use case, provided through the `--config` parameter.
+   
+   The directory also contains a Nextflow configuration file (`nextflow.config`) to fine-tune executor resources and *ksrates* container settings. If needed, please adapt it according to the availbale resources; see the [full documentation](https://ksrates.readthedocs.io/) and the [Nextflow documentation](https://www.nextflow.io/docs/latest/config.html) for more detail on Nextflow configuration,  ...
```
```suggestion
   The `test` directory also already contains a Nextflow configuration file template(`nextflow.config`) to configure the executor to be used (i.e., a local computer or a compute cluster) and its resources made available to Nextflow such as the number of CPUs. It also configures whether to use the *ksrates* Singularity or Docker container. This may need to be adapted to your available resources.
   See the [full documentation](https://ksrates.readthedocs.io/) and the [Nextflow documentation](https://www.nextflow.io/docs/latest/config.html) for more detail on Nextflow configuration, e.g. for different HPC schedulers. We also provide additional, more general template Nextflow configuration files in the [doc](doc/source) directory in the repository.
   If the Nextflow configuration file is named `nextflow.config` the file is automatically detected. Alternatively, the user can specify a custom file by using the `-C` option (see `Nextflow documentation <https://www.nextflow.io/docs/latest/cli.html#hard-configuration-override>`__).
```

lohausr replied:
I think this would fit better in a "Prepare the config files section" as you had before.

lohausr reviewed (changes requested):
See my comments and suggestions. In particular note the comments on the nextflow.config, "-with-singularity" and NXF_SINGULARITY_CACHEDIR issue as also commented on in Campfire.

Cecilia-Sensalari commented on doc/source/usage.rst:
```diff
@@ -34,11 +34,11 @@ The *ksrates* pipeline can be automatically run through Nextflow with a few prep
+    The *ksrates configuration file* is provided through the ``--config`` option. The *Nextflow configuration file* is automatically recognized when it's named with the Nextflow-reserved ``nextflow.config`` filename and it's located in the launching directory; alternatively, the user can provide a costum filename or path through the ``-c`` option.
     
-    The first time the command is launched it downloads the *ksrates* Nextflow pipeline from the ``VIB-PSB/ksrates`` GitHub repository; from then on it uses the local copy stored in the ``.nextflow`` directory. If running a container, the image is pulled from Docker Hub and stored locally for successive usage.  
+    The first time the command is launched it downloads the *ksrates* Nextflow pipeline from the ``VIB-PSB/ksrates`` GitHub repository; from then on it uses the local copy stored in the ``.nextflow`` directory. If running a container, the image is pulled from Docker Hub and stored locally for successive usage.
```
Yes, it's in home, just checked. I'll add it!

Cecilia-Sensalari commented on main.nf:
```diff
@@ -6,6 +6,9 @@ LOG = true  // should probably use our own Logger...
  */
 
+// Parameter to automatically delete or not leftover folders at the end of the pipeline
+params.preserve = false
```
Yes, it's already documented in the master branch and it appears in [RtDs](https://ksrates.readthedocs.io/en/latest/configuration.html#nextflow-configuration-file).
Good idea to add it to the parameters under the NF header!

Cecilia-Sensalari replied:
Hmm... I've added it under the header, but permanently, also when it's false. I don't think it bothers that much...

Cecilia-Sensalari commented on doc/source/installation.rst:
```diff
@@ -138,7 +138,7 @@ Testing your installation
     * Through a Singularity container::
      
-        nextflow run VIB-PSB/ksrates --config ./config_elaeis.txt -with-singularity docker://vibpsb/ksrates:latest
+        nextflow run VIB-PSB/ksrates --config ./config_elaeis.txt -with-singularity vibpsb/ksrates:latest
```
I tried without docker:// and it could pull the container.
I also found in NF [docs](https://www.nextflow.io/docs/latest/singularity.html#singularity-docker-hub) (the second Note paragraph) that "docker://" is not needed anymore when pulling a Singularity container from Docker Hub.
This is actually nice and we could just use "container = 'vibpsb/ksrates:latest'" in the NF config file, instead of having two different lines.
![image](https://user-images.githubusercontent.com/57489957/129178054-a48921d3-c364-4dd8-9f3f-5c9a477e8659.png)

lohausr replied:
Great! Yes, let's just a single `container = 'vibpsb/ksrates:latest'`!

Cecilia-Sensalari commented on test/nextflow.config:
```diff
@@ -0,0 +1,60 @@
+// EXECUTOR SETTINGS:
+executor {
+    name = 'local'              // or e.g. 'sge' when using a computer cluster
+    cpus = 2                    // to limit the CPU resources when using the 'local' executor
```
I've restored 3 and the GitHub runner didn't complain. Let's keep 3 then :)

lohausr commented on README.md:
```diff
@@ -79,18 +79,22 @@ We briefly illustrate here how to run the *ksrates* Nextflow pipeline on the `te
            cd ~/test
 
-2. Launch the *ksrates* Nextflow pipeline.
+2. Prepare the configuration files. The `test` directory already contains:
```
```suggestion
2. Prepare the configuration files.

      The `test` directory already contains:
      
```

lohausr commented on README.md:
```diff
@@ -79,18 +79,22 @@ We briefly illustrate here how to run the *ksrates* Nextflow pipeline on the `te
-   > **Note:** If this is the first time you launch the pipeline, Nextflow will first download *ksrates* and the *ksrates* Singularity container (in the directory specified by `NXF_SINGULARITY_CACHEDIR`, e.g. the launching directory).
+      * A pre-filled *ksrates* configuration file (`config_elaeis.txt`) for the oil palm use case.
+
+      * A Nextflow configuration file template (`nextflow.config`) to configure the executor to be used (i.e., a local computer or a compute cluster) and its resources made available to Nextflow such as the number of CPUs. It also configures whether to use the *ksrates* Singularity or Docker container. This may need to be adapted to your available resources.
```
```suggestion
      * A Nextflow configuration file template (`nextflow.config`) to configure the executor to be used (i.e., a local computer or a compute cluster) and its resources made available to Nextflow such as the number of CPUs. It also configures whether to use the *ksrates* Singularity or Docker container. The configuration file may need to be adapted to your available resources.
```

lohausr commented on README.md:
```diff
@@ -79,18 +79,22 @@ We briefly illustrate here how to run the *ksrates* Nextflow pipeline on the `te
+
+3. Launch the *ksrates* Nextflow pipeline.
+
+   > **Note:** If this is the first time you launch the pipeline, Nextflow will first download *ksrates* in`$HOME/.nextflow` and the *ksrates* Singularity container in `work/singularity`.
```
```suggestion
   > **Note:** If this is the first time you launch the pipeline, Nextflow will first download *ksrates* Nextflow pipeline and the *ksrates* Singularity or Docker container.
```

lohausr commented on README.md:
```diff
@@ -79,18 +79,22 @@ We briefly illustrate here how to run the *ksrates* Nextflow pipeline on the `te
-   The `test` directory also already contains a Nextflow configuration file template (`nextflow.config`) to configure the executor to be used (i.e., a local computer or a compute cluster) and its resources made available to Nextflow such as the number of CPUs. It also configures whether to use the *ksrates* Singularity or Docker container. This may need to be adapted to your available resources.
-   See the [full documentation](https://ksrates.readthedocs.io/) and the [Nextflow documentation](https://www.nextflow.io/docs/latest/config.html) for more detail on Nextflow configuration, e.g. for different HPC schedulers. We also provide additional, more general template Nextflow configuration files in the [doc](doc/source) directory in the repository.
-   If the Nextflow configuration file is named `nextflow.config` the file is automatically detected. Alternatively, the user can specify a custom file by using the `-C` option (see `Nextflow documentation <https://www.nextflow.io/docs/latest/cli.html#hard-configuration-override>`__).
+   The path to the *ksrates* configuration file is specified through the `--config` parameter, while as long as the Nextflow configuration file is named `nextflow.config` and located in the launching folder the file is automatically detected. Alternatively, the user can specify a custom file by using the `-C` option (see `Nextflow documentation <https://www.nextflow.io/docs/latest/cli.html#hard-c ...
```
```suggestion
   The path to the *ksrates* configuration file is specified through the `--config` parameter. If the Nextflow configuration file is named `nextflow.config` and located in the launching folder the file is automatically detected. Alternatively, the user can specify a custom file by using the `-C` option (see `Nextflow documentation <https://www.nextflow.io/docs/latest/cli.html#hard-configuration-override>`__).
```

lohausr commented on doc/source/template_docker_nextflow.config:
```diff
@@ -15,8 +15,7 @@ executor {
 process {
-
-    container = 'vibpsb/ksrates:latest'
+    container = 'vibpsb/ksrates:latest'        // to pull the ksrates container
```
```suggestion
    container = 'vibpsb/ksrates:latest'        // to use the ksrates container
```

lohausr commented on doc/source/template_singularity_nextflow.config:
```diff
@@ -17,7 +17,7 @@ executor {
 
 process {
-    container = 'docker://vibpsb/ksrates:latest'
+    container = 'vibpsb/ksrates:latest'        // to pull the ksrates container
```
```suggestion
    container = 'vibpsb/ksrates:latest'        // to use the ksrates container
```

lohausr commented on example/nextflow.config:
```diff
@@ -20,9 +20,7 @@ executor {
-    // Choose a container type (not needed if all dependencies are locally installed)
-    container = 'docker://vibpsb/ksrates:latest'        // to use the ksrates Singularity container
-    // container = 'vibpsb/ksrates:latest'              // to use the ksrates Docker container
+    container = 'vibpsb/ksrates:latest'        // to pull the ksrates container
```
```suggestion
    container = 'vibpsb/ksrates:latest'        // to use the ksrates container (not needed if all dependencies are locally installed)
```

lohausr commented on test/nextflow.config:
```diff
@@ -15,13 +15,11 @@ singularity {
-    // Choose a container type (not needed if all dependencies are locally installed)
-    container = 'docker://vibpsb/ksrates:latest'        // to use the ksrates Singularity container
-    // container = 'vibpsb/ksrates:latest'              // to use the ksrates Docker container
+    container = 'vibpsb/ksrates:latest'        // to pull the ksrates container
```
```suggestion
    container = 'vibpsb/ksrates:latest'        // to use the ksrates container (not needed if all dependencies are locally installed)
```

lohausr reviewed (changes requested):
A few more small changes.

There are also a few more places in other files where we still describe `-c ./custom_nextflow.config` and `-with-singularity vibpsb/ksrates:latest`. Maybe do a search over all files.

It may also be good to mention the nextflow.config file in the "Testing your installation" section in installation.rst, and that it may need to be adapted.

Cecilia-Sensalari commented:
Removed the last `-c ./custom_nextflow.config` and `-with-singularity vibpsb/ksrates:latest`.
Removed in `installation.rst` the -with-container commands and mentioned nextflow.config and how to use it.

lohausr commented on doc/source/installation.rst:
```diff
@@ -134,16 +134,8 @@ Testing your installation
-    
         nextflow run VIB-PSB/ksrates --config ./config_elaeis.txt
+
+    Nextflow will download *ksrates* and will by default run the test pipeline on the local executor by means of a Singularity container, as configured in the ``nextflow.config`` Nextflow configuration file (automatically detected). If needed, please adapt the configuration to the available resources (e.g. switching to a Docker container) as described in the :ref:`nextflow_config_section` section ...
```
```suggestion
    Nextflow will download *ksrates* and will by default run the test pipeline on the ``local`` executor using the *ksrates* Singularity container, as configured in the included ``nextflow.config`` Nextflow configuration file (automatically detected). If needed, please adapt the configuration to the available resources (e.g. available CPUs/cores or switching to a Docker container or no container at all for a local installation) as described in the :ref:`nextflow_config_section` section.
```

lohausr reviewed (commented):
One last (I think) small text suggestion ;)
